### PR TITLE
style: stricter ruff linters

### PR DESCRIPTION
**Pull request by lengau**
- [x] Have you followed the guidelines for contributing?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `tox`?

-----

Please raise any opinions here. If we don't have unanimity on rules we shouldn't include them.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -166,54 +166,60 @@ select = [  # Base linting rule selections.
     "UP",  # Pyupgrade - note that some of are excluded below due to Python versions
     "YTT",  # flake8-2020: Misuse of `sys.version` and `sys.version_info`
     "ANN",  # Type annotations.
+    "ASYNC",  # Catching blocking calls in async functions
```
Comment (lengau) [inline pyproject.toml:169]: We don't use async anything right now, so this is only future-proofing.

File: pyproject.toml [hunk 2]
```diff
@@ -166,54 +166,60 @@ select = [  # Base linting rule selections.
...
+    # flake8-bandit: security testing. https://docs.astral.sh/ruff/rules/#flake8-bandit-s
```
Comment (lengau) [inline pyproject.toml:170]: This is just moved from below.

File: pyproject.toml [hunk 3]
```diff
@@ -166,54 +166,60 @@ select = [  # Base linting rule selections.
...
+    "S113",  # Requests calls without timeouts
+    "S3",  # Serialising, deserialising, hashing, crypto, etc.
+    "S5",  # Unsafe cryptography or YAML loading.
+    "S602",  # Subprocess call with shell=true
+    "S701",  # jinja2 templates without autoescape
     "BLE",  # Do not catch blind exceptions
     "FBT",  # Disallow boolean positional arguments (make them keyword-only)
     "B0",  # Common mistakes and typos.
     "A",  # Shadowing built-ins.
+    "COM",  # Trailing commas
```
Comment (lengau) [inline pyproject.toml:186]: This was something we previously had enabled in pylint that now has options in ruff.

**Review by tigarmo — APPROVED** [review]

File: pyproject.toml [hunk 4]
```diff
@@ -166,54 +166,66 @@ select = [  # Base linting rule selections.
...
     "C4", # Encourage comprehensions, which tend to be faster than alternatives.
     "T10",  # Don't call the debugger in production code
     "ISC",  # Implicit string concatenation that can cause subtle issues
     "ICN",  # Only use common conventions for import aliases.
     "INP",  # Implicit namespace packages
+    # flake8-pie: miscellaneous linters (enabled individually because they're not really related)
```
Comment (lengau) [inline pyproject.toml:192]: These are new, but I don't think they should be controversial.

**Review by syu-w — APPROVED** [review]

**Review by mr-cal — APPROVED** [review]

File: pyproject.toml [hunk 5]
```diff
@@ -166,54 +166,66 @@ select = [  # Base linting rule selections.
...
     "SIM",  # Code simplification
+    "TID",  # Tidy imports
+    # The team have chosen to only use type-checking blocks when necessary to prevent circular imports.
+    # As such, the only enabled type-checking checks are those that warn of an import that needs to be
+    # removed from a type-checking block.
     "TCH004",  # Remove imports from type-checking guard blocks if used at runtime
     "TCH005",  # Delete empty type-checking blocks
     "ARG",  # Unused arguments
     "PTH",  # Migrate to pathlib
+    "FIX",  # All TODOs, FIXMEs, etc. should be turned into issues instead.
```
Comment (syu-w) [inline pyproject.toml:214]: What should we do with this? What I see in the code is something we are not supported yet but also no plan to support.

**Review by sergiusens — APPROVED** [review]

File: pyproject.toml [hunk 6]
```diff
@@ -166,54 +166,66 @@ select = [  # Base linting rule selections.
...
-    "S506",  # Unsafe YAML load
-    "S508", "S509",  # Insecure SNMP
-    "S701",  # jinja2 templates without autoescape
+    "FLY",  # Detect things that would be better as f-strings.
+    "PERF",  # Catch things that can slow down the application like unnecessary casts to list.
     "RUF001", "RUF002", "RUF003",  # Ambiguous unicode characters
     "RUF005",  # Encourages unpacking rather than concatenation
     "RUF008",  # Do not use mutable default values for dataclass attributes
+    "RUF011",  # Don't use static keys in dict comprehensions.
+    "RUF012",  # Mutable class attributes should be annotated with typing.ClassVar
```
Comment (syu-w) [inline pyproject.toml:225]: Follow RUF012 will broken `pydantic` that it cannot found the `ClassVar`

Comment (lengau) [reply pyproject.toml:225]: Thanks for pointing this out. Removed!

**Pull request merged**